SLING-8706 - Injections for java.util.Optional<> should be automatically optional

etugarev opened the pull request:
Added java.util.Optional<> support

See: https://issues.apache.org/jira/browse/SLING-8706

paul-bjorkstrand commented:
My concern with this impl that it only applies to `@ValueMapValue` injections, and not other injectors (`@ChildResource` or `@RequestAttribute` to name a couple, not to mention the custom ones out there). I wonder if there is a way to make `Optional<..>` work for all injectors, without the need to modify each individual injector.

It will likely require a change to `ModelAdapterFactory.internalCreateModel(..)` that detects `Optional<..>` and does a recursive call. with the type extracted from the field's generic params.

paul-bjorkstrand commented on src/main/java/org/apache/sling/models/impl/injectors/ValueMapInjector.java:
```diff
@@ -98,6 +99,11 @@ public Object getValue(@NotNull Object adaptable, String name, @NotNull Type typ
             Class<?> collectionType = (Class<?>) pType.getRawType();
+
+            if (collectionType.equals(Optional.class)) {
+                return Optional.ofNullable(map.get(name));
```
In addition to [my other concern], this call to `map.get(..)` would need to use the actual type extracted from the `ParameterizedType` of the field, similar to how it is done on [line 70 in this PR]. You can see the extraction of the actual type argument in lines [95][line 95 in this PR]-101 in this file in this PR, or starting on [line 1004][line 1004 of ModelAdapterFactory].

This implementation would likely fail for situations where the type in the JCR does not match the expected injection site type. Example: if you have a `String` field in the model, but the value stored in the JCR is `String[]` or `Long`. Using just `map.get(name)` returns the actual type stored as `Object`. If that is a `String[]`, you cannot (in java) cast to `String`.

[my other concern]: https://github.com/apache/sling-org-apache-sling-models-impl/pull/17#issuecomment-554768731
[line 70 in this PR]: https://github.com/apache/sling-org-apache-sling-models-impl/pull/17/files#diff-c249fe75556c30b251535a7eb4309b7dR70
[line 95 in this PR]: https://github.com/apache/sling-org-apache-sling-models-impl/pull/17/files#diff-c249fe75556c30b251535a7eb4309b7dR95
[line 1004 of ModelAdapterFactory]: https://github.com/apache/sling-org-apache-sling-models-impl/blob/master/src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java#L1004

etugarev replied:
I like the idea to make it work with other injectors too, so I'll give it a shot. Thanks for pointing out the second use case, somehow i missed that. Will get back with a better code!

etugarev commented:
Posting here a "work in progress". 
As discussed I before have modified a `ModelAdapterFactory.injectElement(..)`to incorporate the `Optional<>` class. I changed `ModelAdapterFactory.injectElement(..)` to be a facade and moved its original logic to `ModelAdapterFactory.injectElementInternal(..)`.
 An additional method `getElement()` was created in order "represent" an optional `InjectableElement`. Also a specific version of `injectElementInternal()` is in place to fulfil the injection. Few questions regarding this change:

1. Does it look legitimate?
2. How to deal with the presence of `@Optional` annotation if field is marked as Optional<>?

etugarev commented:
I improved test coverage and tested also on sandbox project, to make sure all the injector-specific annotations are working as before (`@ChildResource`, `@RequestAttribute`, `@ScriptVariable` etc). No problems so far.

justinedelson commented on src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:
```diff
@@ -495,12 +480,97 @@ public RuntimeException inject(InjectableElement element, Object value) {
     }
 
+    @NotNull
+    private InjectableElement getElement(InjectableElement element, Type type) {
```
IIUC, this method is not named intuitively. It should be something like `getOptionalTypeElement` or something like that.

It also might be best to move this logic to the constructor of `InjectableField` or at least move this anonymous inner class to a new class named `OptionalTypedInjectableField` (or something like that since `Optional` is now a bit ambiguous).

etugarev replied:
Thanks for this idea, will give it a try

etugarev replied:
Moved implementation to a named class `OptionalTypedInjectableElement`

paul-bjorkstrand commented on src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:
```diff
@@ -495,12 +480,97 @@ public RuntimeException inject(InjectableElement element, Object value) {
+                    return injectElementInternal(el, adaptable, registry, new InjectCallback() {
+                        @Override
+                        public RuntimeException inject(InjectableElement element1, Object value) {
+                            return callback.inject(element, value.equals(Optional.empty())
+                                    ? value
+                                    : Optional.of(value));
```
This part is perplexing me. Why would you wrap the value in `Optional.of(..)` only when the parameter `value` is not an empty optional?.

To me, this seems like it should be more like this:
```suggestion
                            return callback.inject(element1, Optional.ofNullable(value));
```

paul-bjorkstrand replied:
If you wanted to go a step further, you could replace the entire `InjectCallback` here with a new type, like the ones above in this file:

```java
    private class OptionalWrappingCallbackimplements InjectCallback {

        private final InjectCallback chainedCallback;

        private OptionalWrappingCallback(InjectCallback chainedCallback) {
            this.chainedCallback= chainedCallback;
        }

        @Override
        public RuntimeException inject(InjectableElement element, Object value) {
            return chainedCallback.inject(element, Optional.ofNullable(value));
        }
    }
```

This makes the call to `injectInternal(..)` simpler:

```java
return injectElementInternal(el, adaptable, registry, new OptionalWrappingCallback(callback), preparedValues, modelContext);
```

(Note: I wrote these directly in this review, so forgive any syntax errors).

paul-bjorkstrand replied:
Looks like @justinedelson beat me to the punch with the class suggestion 🙂

etugarev replied:
Moved implementation to `OptionalWrappingCallback` private class.

etugarev replied:
Having `Optional.ofNullable(value)` would be more elegant, however, but will not help as the injection will fail already before this point if `value == null`, so this logic actually to handle the null value properly, please see my comment below.

paul-bjorkstrand commented on src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:
```diff
@@ -495,12 +480,97 @@ public RuntimeException inject(InjectableElement element, Object value) {
+
+            @Override
+            public Object getDefaultValue() {
+                return element.getDefaultValue() == null ? Optional.empty() : element.getDefaultValue();
```
Ok, after reading through this again (after making the comment below), I understand where the confusion comes from. Changing the type here is causing you to have some perplexing code in the `InjectCallback` below. Instead of handling Optional in two places, let the default value propagate `null` and use the null-safe wrapper method, `Optional.ofNullable(..)`.

etugarev replied:
Good idea, this simplifies things a bit, will give it a try.

etugarev replied:
Here, unfortunately, I have to provide a value which is not null, to make the injection successful as we don't have a specific injector for Optional type and a default value injection will be executed last and and will fail if the value is not present (null)

henrykuijpers commented:
Is anything still needed to merge this in, @justinedelson ?

joerghoh commented:
Is there any topic still open? I would then move this topic forward and merge it later this week.

joerghoh commented:
@justinedelson @paul-bjorkstrand: Any objection?
